### PR TITLE
Jasper mask/export fix

### PR DESCRIPTION
**Pull request by borisfom**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/modules/conv_asr.py [hunk 1]
```diff
@@ -69,15 +69,13 @@ def _prepare_for_export(self, **kwargs):
             if isinstance(m, MaskedConv1d):
                 m.use_mask = False
                 m_count += 1
-
-            if isinstance(m, MaskedConv1d):
                 if m.conv.stride[0] > 1 and 'mconv' in name:
                     stride = stride * m.conv.stride[0]
 
-            if isinstance(m, SqueezeExcite):
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:77]: This is necessary. Masked SE will not work without this when exported.

File: nemo/collections/asr/modules/conv_asr.py [hunk 2]
```diff
@@ -241,10 +239,6 @@ def _prepare_for_export(self):
             if isinstance(m, MaskedConv1d):
                 m.use_mask = False
                 m_count += 1
-
-            if isinstance(m, SqueezeExcite):
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:245]: Same as above.

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 3]
```diff
@@ -462,72 +432,66 @@ def __init__(
...
-            y = self._se_pool_step(x, mask)  # [B, C, 1]
-            y = y.transpose(1, -1)  # [B, 1, C]
-            y = self.fc(y)  # [B, 1, C]
-            y = y.transpose(1, -1)  # [B, C, 1]
+            # masked_fill needs to be here because of TRT issue with Where()
+            x = x.float().masked_fill(mask, 0.0)  # mask padded values explicitly to 0
+            y = self._se_pool_step(x, mask).to(x.dtype)
+
+        y = y.transpose(1, -1)  # [B, 1, C]
+        y = self.fc(y)  # [B, 1, C]
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/jasper.py:452]: These 3 ops need to be in fp32 to be stable for SE. Otherwise FP16 causes nans during training.

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 4]
```diff
@@ -462,72 +432,66 @@ def __init__(
...
-        return y, lengths
 
+        return y, lengths
+    
     def _se_pool_step(self, x, mask):
         # Negate mask back to represent 1 for signal and 0 for padded timestep.
         mask = ~mask
-
-        if self.context_window < 0:
+        if self.context_window < 0 or not self.training:
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/jasper.py:467]: Dont merge the two functionalities. I wanted two code paths for unlimited context window and limited context window, so that I can experiment with masked SE eval mode. Please revert this or not self.training. Mask has to be applied during both training and eval, and depends on context window.

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 5]
```diff
@@ -462,72 +432,66 @@ def __init__(
...
+            y = torch.sum(x, dim=-1, keepdim=True).div(mask.sum(dim=-1, keepdim=True).to(x.dtype))
         else:
             # [B, C, 1] - Masked Average over value + padding with limited context.
             # During training randomly subsegments a context_window chunk of timesteps.
             # During inference selects only the first context_window chunk of timesteps.
-            if self.training:
-                y = _se_pool_step_script_train(x, self.context_window, mask)
-            else:
-                y = _se_pool_step_script_infer(x, self.context_window, mask)
+            y = _se_pool_step_script_train(x, self.context_window, mask)
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/jasper.py:474]: Revert, this needs 2 code paths for experimentation.

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 6]
```diff
@@ -169,36 +169,6 @@ def get_asymtric_padding(kernel_size, stride, dilation, future_context):
 
     return (left_context, right_context)
 
-
-@torch.jit.script
-def _se_pool_step_script_infer(x: torch.Tensor, context_window: int, mask: torch.Tensor):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/jasper.py:174]: Revert, i need both code paths for experimentation.

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 7]
```diff
@@ -462,72 +432,66 @@ def __init__(
...
+            self.register_buffer('seq_range', seq_range, persistent=False)
 
-    def make_pad_mask(self, seq_lens, max_audio_length, device=None):
+    def make_pad_mask(self, seq_lens, max_audio_length):
         """Make masking for padding."""
-        if device and self.seq_range.device != device:
-            self.seq_range = self.seq_range.to(device)
-
-        if self.seq_range.device != seq_lens.device:
-            seq_lens = seq_lens.to(self.seq_range.device)
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/jasper.py:524]: These two lines are necessary for inference without Pytorch Lightning - there is no guarantee that the tensor are on correct device when you call model.transcribe() or other high level apis that don't depend on PTL

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 8]
```diff
@@ -462,72 +432,66 @@ def __init__(
...
-
-        mask = self.seq_range[:max_audio_length].expand(seq_lens.size(0), -1) < seq_lens.unsqueeze(-1)  # [B, T]; bool
+        mask = self.seq_range[:max_audio_length].repeat(seq_lens.size(0),1) < seq_lens.unsqueeze(-1)  # [B, T]; bool
         mask = mask.unsqueeze(1)  # [B, 1, T]
-
-        if device and mask.device != device:
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/jasper.py:529]: What is the harm in this? It moves the mask only if device doesn't match.

**Review by titu1994 — COMMENTED** [review]
initial comments

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 9]
```diff
@@ -462,72 +432,66 @@ def __init__(
...
+    def make_pad_mask(self, seq_lens, max_audio_length):
         """Make masking for padding."""
-        if device and self.seq_range.device != device:
-            self.seq_range = self.seq_range.to(device)
-
-        if self.seq_range.device != seq_lens.device:
-            seq_lens = seq_lens.to(self.seq_range.device)
-
-        mask = self.seq_range[:max_audio_length].expand(seq_lens.size(0), -1) < seq_lens.unsqueeze(-1)  # [B, T]; bool
+        mask = self.seq_range[:max_audio_length].repeat(seq_lens.size(0),1) < seq_lens.unsqueeze(-1)  # [B, T]; bool
```
[the remark this answers is not in the document]
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/jasper.py:493]: No bug, but for the same code in conformer I found repeat generated better ONNX code.

File: nemo/collections/asr/modules/conv_asr.py (hunk 1 above)
Comment (borisfom) [reply nemo/collections/asr/modules/conv_asr.py:77]: Why, if the masks are taken care of from forward() ?

File: nemo/collections/asr/modules/conv_asr.py (hunk 2 above)
Comment (borisfom) [reply nemo/collections/asr/modules/conv_asr.py:245]: Why, if the masks are taken care of from forward() ?

File: nemo/collections/asr/parts/submodules/jasper.py (hunk 8 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/jasper.py:529]: That case would only happen if two inputs (x, length) are on different devices - this function only is called with device=x.device. Is that real use case ?

File: nemo/collections/asr/modules/conv_asr.py [hunk 10]
```diff
@@ -218,16 +208,27 @@ def update_max_sequence_length(self, seq_length: int, device):
...
+            else:
+                self.register_buffer('seq_range', seq_range, persistent=False)
 
             # Update all submodules
             for name, m in self.named_modules():
                 if isinstance(m, MaskedConv1d):
-                    if m.use_mask:
-                        m.update_masked_length(self.max_audio_length, device=device)
-
+                    m.max_len = self.max_audio_length
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:227]: Can we update the jasper method update_masked_length() to take these two values instead of access inner members like this? Or even have a separate function to update these lengths in place without allocating.

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 11]
```diff
@@ -462,17 +462,21 @@ def __init__(
...
+        self.set_max_len(16)
 
     def forward(self, x, lengths):
         return self.forward_for_export(x, lengths)
 
     def forward_for_export(self, x, lengths):
         # The use of negative indices on the transpose allow for expanded SqueezeExcite
+        max_len = x.shape[-1]
+        if max_len > self.max_len:
+            self.set_max_len(max_len)
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/jasper.py:474]: Is this fine inside of export? Wont the calculation of new mask affect ONNX?

File: nemo/collections/asr/parts/submodules/jasper.py [hunk 12]
```diff
@@ -481,13 +485,13 @@ def forward_for_export(self, x, lengths):
...
+            # Note: Keep for future, in case we improve WER from doing so.
+            # if self.context_window >= 0:
+            #     y = F.interpolate(y, size=x.shape[-1], mode=self.interpolation_mode)
 
-        y = torch.sigmoid(y)
-        y = x * y
-        return y, lengths
+            y = torch.sigmoid(y)
+            y = x * y
+        return y.to(dtype=dtype), lengths
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/jasper.py:494]: The output of SE block must be fp32. We see some instability in fp16 training otherwise.

File: nemo/collections/asr/parts/submodules/jasper.py (hunk 7 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/jasper.py:524]: ^

File: nemo/collections/asr/parts/submodules/jasper.py (hunk 8 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/jasper.py:529]: When we do manual transcription, this can occur. Same with above, when PTL isn't handling it there were issues thas why I added these lines.

**Review by titu1994 — COMMENTED** [review]
Overall looks fine but some changes necessary

File: nemo/collections/asr/parts/submodules/jasper.py (hunk 11 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/jasper.py:474]: This will be traced and if() condition will be false, so no-op.

File: nemo/collections/asr/modules/conv_asr.py (hunk 10 above)
Comment (borisfom) [reply nemo/collections/asr/modules/conv_asr.py]: done.

File: nemo/collections/asr/parts/submodules/jasper.py (hunk 12 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/jasper.py]: done.

File: nemo/collections/asr/parts/submodules/jasper.py (hunk 8 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/jasper.py:529]: Restored. The third device if() - literally cannot occur.

**Review by titu1994 — APPROVED** [review]
Lgtm
